### ISSUE TITLE
Pushing stream to remote using RESTAPI crashes server

### ISSUE THREAD

**Issue by artasoftware** [AirenSoft/OvenMediaEngine#331]
Hello.
I'm trying to restream using RTMPPush module. 

I've enabled RTMPPush, enabled API and sent this POST request using curl:

curl -d '{"stream":"stream.url.com","name":"stream_name","protocol":"rtmp","url":"rtmp://stream.url.com:1935/live/","streamKey":"stream_name"}' -H 'Authorization: Basic b21lLWFjY2Vzcy10b2tlbg==' -H 'Content-Type: application/json' http://127.0.0.1:48801/v1/vhosts/default/apps/app:startPush

And receiving this response: 
curl: (52) Empty reply from server

Log contains this strings: 

[2021-04-07 22:10:27.586] I [SckPoolAPIServe:30356] Socket.Server | server_socket.cpp:129  | [#7] [0x2255f80] Client(<ClientSocket: 0x7f5c5c000970, #71, state: Closed, TCP, 127.0.0.1:44178>) is disconnected
[2021-04-07 22:12:02.060] C [SckPoolAPIServe:30356] OvenMediaEngine | signals.cpp:120  | OME received signal 6 (SIGABRT), interrupt.

And server stops. 

I've attached my crash dump and Server.xml.  

Besides, I should mention that several parameters I've used in request are similar (my true values replaced for security reasons because everything I was provided with to push stream was two strings: "rtmp://stream.url.com:1935/live/" and "streamname: stream_name". 
So I've bolded similar parts I've used:
"stream" = "**stream.url.com**" and "url":"rtmp://**stream.url.com**:1935/live/" 
"name":"**stream_name**" and "streamKey":"**stream_name**"

Thanks in advance. 

[ovenmedia.zip](https://github.com/AirenSoft/OvenMediaEngine/files/6272907/ovenmedia.zip)

**Comment by kwonkeukhan** [comment]
Hi. @artasoftware 

The server crashed due to an error in parsing the wrong json document.  I patched the latest git master.

And, the json request you made is wrong. Please refer to the code below.

`{
  "id" : "UNIQUE_ID",
  "stream": {
    "name": "OUTPUT_STREAM_NAME_OF_OME"
  },
  "protocol": "rtmp",
  "url":"rtmp://stream.url.com:1935/live",
  "steramkey":"DESTINATION_STREAM_NAME"
}`

rtmppush is a beta feature. If you have any features you need, please give us your opinion.

Best regards.

**Comment by artasoftware** [comment]
Thanks for the reply.
If I correctly understood: I'm sending stream (usign OBS for example) to my server using rtmp://my.server.com:1935/app, streamname: stream (all variables are default). Simultaneously to this I'm starting RTMPPush with this request:

curl -d '{ "id" : "1", "stream": { "name": "stream" }, "protocol": "rtmp", "url":"rtmp://stream.url.com:1935/live", "steramkey":"stream_key" }' -H 'Authorization: Basic b21lLWFjY2Vzcy10b2tlbg==' -H 'Content-Type: application/json' http://127.0.0.1:48081/v1/vhosts/default/apps/app:startPush

If above is correct then I've got something wrong because "pushes" returns all zeros:
         "totalsentTime" : 0,
         "totalsentBytes" : 0,
         "sentBytes" : 0,
         "sentTime" : 0,

Log contains these strings (tried various push urls):

[2021-04-09 19:24:47.348] W [RtmpPubWorker:8971] RTMPPush | rtmppush_session.cpp:70   | Could not supported codec. codec_id(8)
[2021-04-09 19:24:47.348] E [RtmpPubWorker:8971] RTMPWriter | rtmp_writer.cpp:137  | Error opening file. error(-1330794744:Protocol not found), filename(stream.url.com:1935/live/)
[2021-04-09 19:24:47.348] W [RtmpPubWorker:8971] RTMPPush | rtmppush_session.cpp:70   | Could not supported codec. codec_id(8)
[2021-04-09 19:24:47.348] E [RtmpPubWorker:8971] RTMPWriter | rtmp_writer.cpp:137  | Error opening file. error(-1330794744:Protocol not found), filename(stream.url.com:1935/live/stream_key/)
[2021-04-09 19:24:47.348] E [StreamWorker:13528] RTMPPush | rtmppush_session.cpp:152  | Failed to add packet
[2021-04-09 19:24:47.348] C [StreamWorker:13528] OvenMediaEngine | signals.cpp:120  | OME received signal 11 (SIGSEGV), interrupt.
[2021-04-09 19:24:47.348] W [RtmpPubWorker:8971] RTMPPush | rtmppush_session.cpp:70   | Could not supported codec. codec_id(8)
[2021-04-09 19:24:47.349] E [RtmpPubWorker:8971] RTMPWriter | rtmp_writer.cpp:137  | Error opening file. error(-1330794744:Protocol not found), filename(stream.url.com:1935/live/)

**Comment by kwonkeukhan** [comment]
There was a typo in the first comment. Please check the revised comments.

steramKey -> streamKey

Thank you.

**Comment by artasoftware** [comment]
Funny that I've noticed that typo but didn't correct it when were sending requests to RESTAPI.
So anyway that didn't make sense. Log attached. 
[ovenmediaengine.zip](https://github.com/AirenSoft/OvenMediaEngine/files/6287418/ovenmediaengine.zip)

**Comment by getroot** [comment]
@artasoftware 

If you've run prerequisites.sh for a long time, or don't prefix your "url" with "rtmp://" in the json, it could be.

**Comment by artasoftware** [comment]
> 
> 
> @artasoftware
> 
> If you've run prerequisites.sh for a long time, or don't prefix your "url" with "rtmp://" in the json, it could be.

I've run it may be a week ago but done it again. My push url is without "rtmp://" for now but result is the same.

**Comment by kwonkeukhan** [comment]
@artasoftware 

When I checked the log you sent, there was a "/" character at the end of the url. Remove the "/" character or download the latest source code. I just added exception handling.
Finally... you need to prefix the url parameter with "rtmp://"

**Comment by artasoftware** [comment]
> 
> 
> @artasoftware
> 
> When I checked the log you sent, there was a "/" character at the end of the url. Remove the "/" character or download the latest source code. I just added exception handling.
> Finally... you need to prefix the url parameter with "rtmp://"

New logs attached. As for "/" at the end of the url - I don't use it when create push stream with RESTAPI, but I confirm it persists in logs. My current request to RESTAPI is:
curl -d '{ "id" : "1", "stream": { "name": "stream" }, "protocol": "rtmp", "url":"stream.url.com/live", "streamkey":"cs" }' -H 'Authorization: Basic b21lLWFjY2Vzcy10b2tlbg==
[ovenmediaengine.zip](https://github.com/AirenSoft/OvenMediaEngine/files/6311583/ovenmediaengine.zip)
' -H 'Content-Type: application/json' http://127.0.0.1:48081/v1/vhosts/default/apps/app:startPush

**Comment by getroot** [comment]
1. Your RTMP input is very unstable. As you can see from the next log, the RTMP input keeps repeating the creation/deletion. In this situation, RTMP Push cannot be done normally. First, create a stream by inserting RTMP input, then check if the playback works well with WebRTC or HLS. 

The reason why the RTMP input is unstable is probably due to the very high bit rate. Right now you are broadcasting at 11Mb at 60fps. Does your environment support this well?

```
[2021-04-09 21:35:26.750] I [SckPoolRTMP:14032] MediaRouter | mediarouter_application.cpp:243  | Trying to create a stream: [#default#app/stream(70)]
[2021-04-09 21:35:26.750] I [SckPoolRTMP:14032] MediaRouter | mediarouter_application.cpp:244  | 
[Stream Info]
id(70), output(stream), SourceType(Rtmp), Created Time (Fri Apr  9 21:35:23 2021)

	Video Track #0: Bypass(false) Bitrate(11.00Mb) codec(1, H264) resolution(1920x1080) framerate(60.00fps) timebase(1/1000)
	Audio Track #1: Bypass(false) Bitrate(192.00Kb) codec(6, AAC) samplerate(48.0K) format(s16, 16) channel(stereo, 2) timebase(1/1000)
[2021-04-09 21:35:26.750] I [SckPoolRTMP:14032] MediaRouter | mediarouter_stream.cpp:55   | Trying to create media route stream: name(stream) id(70)
[2021-04-09 21:35:26.750] I [SckPoolRTMP:14032] Monitor | application_metrics.cpp:56   | Create StreamMetrics(stream) for monitoring
[2021-04-09 21:35:26.751] I [SckPoolRTMP:14032] Transcoder | transcode_stream.cpp:492  | [#default#app/stream(70)] -> [#default#app/stream(694190925)] Output stream has been created.
[2021-04-09 21:35:26.751] I [SckPoolRTMP:14032] MediaRouter | mediarouter_application.cpp:243  | Trying to create a stream: [#default#app/stream(694190925)]
[2021-04-09 21:35:26.752] I [SckPoolRTMP:14032] MediaRouter | mediarouter_application.cpp:244  | 
[Stream Info]
id(694190925), output(stream), SourceType(Transcoder), Created Time (Fri Apr  9 21:35:26 2021)
	>> Origin Stream Info
	id(70), output(stream), SourceType(Rtmp), Created Time (Fri Apr  9 21:35:23 2021)

	Video Track #0: Bypass(true) Bitrate(11.00Mb) codec(1, H264) resolution(1920x1080) framerate(60.00fps) timebase(1/1000)
	Audio Track #1: Bypass(true) Bitrate(192.00Kb) codec(6, AAC) samplerate(1.0K) format(s16, 16) channel(stereo, 2) timebase(1/1000)
	Audio Track #2: Bypass(false) Bitrate(128.00Kb) codec(8, OPUS) samplerate(48.0K) format(s16, 16) channel(stereo, 2) timebase(1/48000)
[2021-04-09 21:35:26.752] I [SckPoolRTMP:14032] MediaRouter | mediarouter_stream.cpp:55   | Trying to create media route stream: name(stream) id(694190925)
[2021-04-09 21:35:26.752] I [SckPoolRTMP:14032] Monitor | application_metrics.cpp:56   | Create StreamMetrics(stream) for monitoring
[2021-04-09 21:35:26.752] I [SckPoolRTMP:14032] Transcoder | transcode_stream.cpp:107  | [#default#app/stream(70)] Transcoder input stream has been started. Status : (1) Decoders, (0) Encoders
[2021-04-09 21:35:26.752] W [InboundWorker:14036] MediaRouter | mediarouter_stream.cpp:1040 | Detected abnormal increased pts. track_id : 1, prv_pts : 0, cur_pts : 635856, crt_pts : 635856, avg_inc : 0, inc : 635856
[2021-04-09 21:35:26.752] I [Decoderaac:30675] Transcoder | transcode_codec_dec_aac.cpp:155  | [#default#app/stream(70)] input stream information: [audio] aac (LC), 48000 Hz, stereo, fltp, 190 kbps, timebase: 1/1000, frame_size: 1024
[2021-04-09 21:35:26.948] W [InboundWorker:14036] MediaRouter | mediarouter_stream.cpp:1040 | Detected abnormal increased pts. track_id : 0, prv_pts : 0, cur_pts : 637679, crt_pts : 637679, avg_inc : 0, inc : 637679
[2021-04-09 21:35:26.950] I [OutboundWorker:14039] WebRTC Publisher | rtc_stream.cpp:267  | Unsupported codec(Audio/AAC) is being input from media track
[2021-04-09 21:35:26.951] I [OutboundWorker:14039] Publisher | stream.cpp:181  | WebRTC Publisher Application application has started [stream(694190925)] stream
[2021-04-09 21:35:26.951] I [OutboundWorker:14039] Publisher | stream.cpp:181  | HLS Publisher Application application has started [stream(694190925)] stream
[2021-04-09 21:35:26.951] I [OutboundWorker:14039] Publisher | stream.cpp:181  | DASH Publisher Application application has started [stream(694190925)] stream
[2021-04-09 21:35:26.951] I [OutboundWorker:14039] Publisher | stream.cpp:181  | LLDASH Publisher Application application has started [stream(694190925)] stream
[2021-04-09 21:35:26.951] I [OutboundWorker:14039] Publisher | stream.cpp:181  | OVTPublisher Application application has started [stream(694190925)] stream
[2021-04-09 21:35:26.952] I [OutboundWorker:14039] Publisher | stream.cpp:181  | RTMPPushPublisher Application application has started [stream(694190925)] stream
[2021-04-09 21:35:26.952] I [AppWorker:14045] DASH | dash_packetizer.cpp:231  | [#default#app/stream] DASH: init_video.m4s has created
[2021-04-09 21:35:26.952] I [AppWorker:14045] DASH | dash_packetizer.cpp:265  | [#default#app/stream] DASH: init_audio.m4s has created
[2021-04-09 21:35:37.471] I [AppWorker:14047] LLDASH | cmaf_packetizer.cpp:750  | [0x7f98680111a0] LLDASH segment is ready to stream [#default#app/stream], segment duration: 5.000000s, count: 1
[2021-04-09 21:36:47.667] I [SckPoolRTMP:14032] RTMPProvider | rtmp_provider.cpp:138  | The RTMP client has disconnected: [#default#app/stream], remote: <ClientSocket: 0x7f988001d080, #70, state: Closed, TCP, 147.75.102.61:40584>
[2021-04-09 21:36:47.667] I [SckPoolRTMP:14032] Provider | stream.cpp:55   | #default#app/stream(70) has been stopped playing stream
[2021-04-09 21:36:47.667] I [SckPoolRTMP:14032] MediaRouter | mediarouter_application.cpp:424  | Trying to delete a stream: [#default#app/stream(70)]
[2021-04-09 21:36:47.667] I [SckPoolRTMP:14032] Monitor | application_metrics.cpp:68   | Delete StreamMetrics(stream) for monitoring
[2021-04-09 21:36:47.667] I [SckPoolRTMP:14032] Monitor | stream_metrics.cpp:31   | 
[Stream Info]
id(70), output(stream), SourceType(Rtmp), Created Time (Fri Apr  9 21:35:23 2021)

	Video Track #0: Bypass(false) Bitrate(11.00Mb) codec(1, H264) resolution(1920x1080) framerate(60.00fps) timebase(1/1000)
	Audio Track #1: Bypass(false) Bitrate(192.00Kb) codec(6, AAC) samplerate(48.0K) format(fltp, 32) channel(stereo, 2) timebase(1/1000)

	>> Statistics
	Last update time : Fri Apr  9 21:36:47 2021, Last sent time : Fri Apr  9 21:35:26 2021
	Bytes in : 63.22MB, Bytes out : 0B, Concurrent connections : 0, Max connections : 0 (Fri Apr  9 21:35:26 2021)

		>>>> By publisher
		- Unknown : Bytes out(0B) Concurrent Connections (0)
		- WebRTC : Bytes out(0B) Concurrent Connections (0)
		- RTMP : Bytes out(0B) Concurrent Connections (0)
		- RTMPPush : Bytes out(0B) Concurrent Connections (0)
		- HLS : Bytes out(0B) Concurrent Connections (0)
		- DASH : Bytes out(0B) Concurrent Connections (0)
		- LLDASH : Bytes out(0B) Concurrent Connections (0)
		- OVT : Bytes out(0B) Concurrent Connections (0)
		- File : Bytes out(0B) Concurrent Connections (0)
		- Thumbnail : Bytes out(0B) Concurrent Connections (0)
```

2. And as @kwonkeukhan  said, when calling the API, add the "rtmp://" prefix to the url like below.

`{"id": "1", "stream": {"name": "stream" }, "protocol": "rtmp", "url":"rtmp://stream.url.com/live", " streamkey":"cs" }'`
